pkg/conn/basedb.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conn
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/pkg/retry"
+	"github.com/pingcap/dm/pkg/terror"
+)
+
+// DBProvider providers BaseDB instance
+type DBProvider interface {
+	Apply(config config.DBConfig) (*BaseDB, error)
+}
+
+type defaultDBProvider struct {
+}
+
+// DefaultDBProvider is global instance of DBProvider
+var DefaultDBProvider DBProvider
+
+func init() {
+	DefaultDBProvider = &defaultDBProvider{}
+}
+
+func (d *defaultDBProvider) Apply(config config.DBConfig) (*BaseDB, error) {
+	dsn := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8mb4&interpolateParams=true&maxAllowedPacket=%d",
+		config.User, config.Password, config.Host, config.Port, *config.MaxAllowedPacket)
+
+	maxIdleConn := 0
+	rawCfg := config.RawDBCfg
+	if rawCfg != nil {
+		if rawCfg.ReadTimeout != "" {
+			dsn += fmt.Sprintf("&readTimeout=%s", rawCfg.ReadTimeout)
+		}
+		if rawCfg.WriteTimeout != "" {
+			dsn += fmt.Sprintf("&writeTimeout=%s", rawCfg.WriteTimeout)
+		}
+		maxIdleConn = rawCfg.MaxIdleConns
+	}
+	db, err := sql.Open("mysql", dsn)
+
+	if maxIdleConn > 0 {
+		db.SetMaxIdleConns(maxIdleConn)
```

Comment:
can not set value `< 0`?

Reply:
How about 
```
if value < 0 then do nothing (default 2 idle connection)
if value == 0 then no idle connection
if value > 0 then set value
```

Reply:
fine, add some comments to describe it.

Reply:
Done in [here](https://github.com/pingcap/dm/pull/266/files#diff-184b4272aa0f082766a0d470548f0644R46)
comments is [here](https://github.com/pingcap/dm/pull/266/files#diff-94b1e02e1fe1bd32d9eee7a40a730b00R67)